### PR TITLE
Listview

### PR DESCRIPTION
**Pull request by JoshMarler**
@nick-thompson,

This is approaching done I think. Main questions would be the following:

. I'd like to support `scrollToIndex` and `scrollToItem` functionality. Do you have am opinion on how this should look? I've been looking to ReactNative `<FlatList>` for inspiration. 

. Should we support a horizontal `ListView` or leave this as a future addition?

. Any other API functionality you would like to see here?

### CODE REVIEW

File: blueprint/core/blueprint_ScrollView.h
Comment (JoshMarler) [inline blueprint/core/blueprint_ScrollView.h:12]: Don't think this is needed. Might have been IDE. Will clean.

File: blueprint/core/blueprint_ScrollView.h
Comment (JoshMarler) [inline blueprint/core/blueprint_ScrollView.h:25]: Should be private. Will fix.

**Comment by nick-thompson** [comment]
Super exciting @JoshMarler thanks for this. Lemme read up on RN's flatList and I'll take a detailed pass tomorrow

**Comment by nick-thompson** [comment]
Ok coming back for a more detailed pass– the code here looks excellent, I'm quite ready to merge once we touch up the few things you've already pointed out. As for your bullets above:

> . I'd like to support scrollToIndex and scrollToItem functionality. Do you have am opinion on how this should look? I've been looking to ReactNative <FlatList> for inspiration.

Yea, let's aim for these. As for how it should look: I think functions makes more sense, like the RN API. The idea of updating a prop to force the scroll feels a little awkward since really the goal is just to mutate some internal state (specifically, the scroll position of the underlying viewport). The tricky bit here is that to implement functions like this, we need to implement something like a `ListViewInstance` in the js side of our renderer that implements `scrollToItem`, bind an appropriate C++ method to that instance (or do some ReactAppRoot top-level think for invoking a named method on some viewTable View based on the given ViewId), and then wire it into our `viewRef` prop that we added for supporting React's refs. It's maybe the last real bit of uncharted territory in the project, so I can't speak much to how difficult it will be, but it may be quite non-trivial!

> Should we support a horizontal ListView or leave this as a future addition?

Let's leave that for the future

> Any other API functionality you would like to see here?

No but an example would be huge. I suspect people will very much like this component, given that we do a lot of preset and sample browsing in our corner of software products!

**Comment by JoshMarler** [comment]
>The tricky bit here is that to implement functions like this, we need to implement something like a ListViewInstance in the js side of our renderer that implements scrollToItem, bind an appropriate C++ method to that instance (or do some ReactAppRoot top-level think for invoking a named method on some viewTable View based on the given ViewId), and then wire it into our viewRef prop that we added for supporting React's refs. It's maybe the last real bit of uncharted territory in the project, so I can't speak much to how difficult it will be, but it may be quite non-trivial!

Right yeah OK. I hadn't given thought to this but I see what you're getting at. I'll digest and take a stab at this today. 

Leaving horizontal ListView for the moment is fine with me. 

>No but an example would be huge. I suspect people will very much like this component, given that we do a lot of preset and sample browsing in our corner of software products!

Yep totally. Once I've got this and the other couple of things I need finished I'll try to pick that "Component Showcase" app back up and fire over to you for a demo. We can decide if we're happy with that forming a basic example of all the stock components.

**Comment by JoshMarler** [comment]
@nick-thompson, 

Noting this here for reference and notes for myself as much as anything. 

So we may be able to just switch on the `elementType` string passed to `BlueprintRenderer.createInstance` and if `"ListView"` call through to `BlueprintBackend.createListViewInstance` where `ListViewInstance` inherits `ViewInstance` and adds the additional functions bound to native?

**Comment by nick-thompson** [comment]
> So we may be able to just switch on the elementType string passed to BlueprintRenderer.createInstance and if "ListView" call through to BlueprintBackend.createListViewInstance where ListViewInstance inherits ViewInstance and adds the additional functions bound to native?
>
> Assumption being we then follow roughly the approach of something like setRawTextValue in which a viewId is passed and we lookup to invoke the relevant native binding to juce::Viewport

Man, you're good. Sometimes it feels like you know this codebase better than I do!

Generally: yes, that's spot on. But there's nuance here: the distinction we have between "Views" and "Text Views" is a distinction we can safely rely on because React's reconciler also deals in this distinction. It can do so because in the Web DOM, text nodes are always truly leaf nodes in the tree, which gives the reconciler a small helpful assumption to rely on. Now, apart from View/Text, the reconciler doesn't care to differentiate any View type from any other View type, so whether we render a `ViewInstance` or a `ListViewInstance` inside of `BlueprintBackend` is of no consequence from the reconciler's point of view. The consequence shows up in the way that we communicate from a `ListViewInstance` through to the actual native `ListView` held in the ViewManager's view table. For TextView, we go right through ReactAppRoot where we can lookup the right view based on its ID, and that feels ok to me because the reconciler is careful to make this distinction between text/view nodes too.

*But* I don't know if I like this as a generic approach for tying View instance methods to methods on the native view instance (and imagine, what if I wrote a custom SpectrogramView for my product and I wanted to add a `reset()` method to it or whatever... as an end user, I don't want to fork blueprint to add my own SpectrogramViewInstance into BlueprintBackend and wire the call through ReactAppRoot). So, I wonder, can we tackle a generic approach here?

My initial idea for that is this.. suppose we wrap every `ViewInstance` returned to a user via the `viewRef` in an ES6 `Proxy` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy). Now with the Proxy we can catch every method invocation and ask "Does the `ViewInstance` _itself_ have a method with this name?" If it does, we just invoke it on the view instance, no problem. If it doesn't, perhaps we assume that the user is trying to call a native method on the underlying native view instance. Maybe at that time we ask ReactAppRoot to dynamically invoke a named method on a given view instance with some arbitrary args list?

In detail:
https://github.com/nick-thompson/blueprint/blob/master/packages/react-juce/src/lib/BlueprintBackend.ts#L125
```
    if (propKey === 'viewRef') {
      value.current = new Proxy(this, {
            get: function(target, prop, receiver) {
                if (target.hasOwnProperty(prop))
                    return target[prop];

                // Else the user is looking up something we don't have... let's return a proxy function to native
                return function(...args) {
                    return __BlueprintNative__.invokeViewMethod(target._id, prop, ...args);
                };         
            }
      });
      return;
    }
```

Now in ReactAppRoot, we implement a native "invokeViewMethod" which uses that first argument (viewId) to look up the view, then invokes a method named by `prop` with `...args`.

To do this, I think we probably have to update the idea of `blueprint::View` to "expose" methods to this interface. For example...

```
class ListView : public blueprint::View
{
    ListView()
    {
          export("scrollToIndex", [this](juce::NativeFunctionArgs const& args) {
              scrollToIndex(args[0]); // Or something
          });
    }
}
```

Where `export` really just adds to a `std;;map<juce::String, juce::NativeFunction>` and the `View` class gives ReactAppRoot some way to look stuff up in there and call the funcs.

Whew, that was a lot, let me know if that makes sense!

**Comment by JoshMarler** [comment]
>Man, you're good. Sometimes it feels like you know this codebase better than I do!

Ha! High praise. Thanks but I think I had a cheeky advantage given I wrote the viewRef code!

>But I don't know if I like this as a generic approach for tying View instance methods to methods on the native view instance (and imagine, what if I wrote a custom SpectrogramView for my product and I wanted to add a reset() method to it or whatever... as an end user, I don't want to fork blueprint to add my own SpectrogramViewInstance into BlueprintBackend and wire the call through ReactAppRoot). So, I wonder, can we tackle a generic approach here?

Yeah 100%. This was my one concern as this explodes potentially. 


> If it does, we just invoke it on the view instance, no problem. If it doesn't, perhaps we assume that the user is trying to call a native method on the underlying native view instance. Maybe at that time we ask ReactAppRoot to dynamically invoke a named method on a given view instance with some arbitrary args list?

Yeah awesome. I was thinking about a generic "invoke registered thing" but I would NEVER have known about that proxy object pattern. Gosh. More and more JS foo to learn. That's a really neat trick though. 

OK, awesome @nick-thompson. Thanks for this. As ever direction is much appreciated. I'll get to implementing this today. Seems reasonably straight forward and I much prefer this approach to adding the likes of `ListViewInstance`. 

If we assume the user is trying to invoke a native function when an object property is not present on the underlying `ViewInstance` then hopefully our usual `std::runtime_error` exception propagation is a sensible enough approach for a non-existent "native prop". 

Awesome stuff, opens up as useful pattern for future features.

**Comment by JoshMarler** [comment]
@nick-thompson, OK, I think this is ready for a proper review. I'll try to get this into the showcase/example app when time allows but have put a simple example usage in `ListView`'s doc comment.

File: blueprint/core/blueprint_View.cpp
Comment (nick-thompson) [inline blueprint/core/blueprint_View.cpp:283]: Nit: we don't need the `else` clause here, because we `return` from the `if`. So can just

```
if (it != nativeMethods.end())
    return it->second(args);

throw std::logic_error....
```

File: blueprint/core/blueprint_ViewManager.cpp
Comment (nick-thompson) [inline blueprint/core/blueprint_ViewManager.cpp:234]: Can we throw here and catch it in the js runtime?

File: packages/react-juce/src/components/ListView.tsx
Comment (nick-thompson) [inline packages/react-juce/src/components/ListView.tsx:212]: Major nitpick: camelCase please. Once we get `prettier` involved we'll catch these things automatically

File: blueprint/core/blueprint_ScrollView.h
Comment (nick-thompson) [inline blueprint/core/blueprint_ScrollView.h:32]: Major nitpick: lower case for the method name's first letter please! Clang-format will save us soon

**Review by nick-thompson — APPROVED** [review]
@JoshMarler this is super super awesome. Love the abstraction here, and I love the pattern for native view methods. Tiny nitpicks inline, and I'll also ask to squash it down into one or two commits before merge. Otherwise, good to go!

File: blueprint/core/blueprint_View.cpp
Comment (JoshMarler) [reply blueprint/core/blueprint_View.cpp:283]: Yup no worries. think I'd been watching a talk advocating explicit `else` and got overly inspired!

**Comment by JoshMarler** [comment]
@nick-thompson , should be good for final pass and merge. Have published 0.2.7 package.

**Comment by nick-thompson** [comment]
🤘 ! Thank you!

**Pull request merged**
